pallets/asset-handler/src/lib.rs

```diff
@@ -0,0 +1,468 @@
+		ConversionIssue,
+		/// BridgeDeactivated
+		BridgeDeactivated,
+		/// Whitelisted token limit reached
+		WhitelistedTokenLimitReached,
+		/// This token is not Whitelisted
+		TokenNotWhitelisted,
+		/// This token was whitelisted but got removed and is not valid anymore
+		WhitelistedTokenRemoved,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		/// On Initialize
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let withdrawal_execution_block =
+				n.saturating_sub(<WithdrawalExecutionBlockDiff<T>>::get());
+			if !withdrawal_execution_block.is_zero() {
+				let pending_withdrawals = <PendingWithdrawals<T>>::get(withdrawal_execution_block);
+				for withdrawal in pending_withdrawals {
+					if chainbridge::Pallet::<T>::transfer_fungible(
+						withdrawal.chain_id,
+						withdrawal.rid,
+						withdrawal.recipient.0.to_vec(),
+						Self::convert_balance_to_eth_type(withdrawal.amount),
+					)
+					.is_err()
+					{
+						Self::deposit_event(Event::<T>::FungibleTransferFailed);
+					}
+				}
+			}
+			// TODO: Benchmark on initialize
+			(195_000_000 as Weight)
+				.saturating_add(T::DbWeight::get().writes(5 as Weight))
+				.saturating_add(T::DbWeight::get().reads(5 as Weight))
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Creates new Asset where AssetId is derived from chain_id and contract Address
+		///
+		/// # Parameters
+		///
+		/// * `origin`: `Asset` owner
+		/// * `chain_id`: Asset's native chain
+		/// * `contract_add`: Asset's actual address at native chain
+		#[pallet::weight(T::WeightInfo::create_asset(1))]
+		pub fn create_asset(
+			origin: OriginFor<T>,
+			chain_id: BridgeChainId,
+			contract_add: H160,
+		) -> DispatchResult {
+			T::AssetCreateUpdateOrigin::ensure_origin(origin)?;
+			let rid = chainbridge::derive_resource_id(chain_id, &contract_add.0);
+			T::AssetManager::create(
+				Self::convert_asset_id(rid),
+				chainbridge::Pallet::<T>::account_id(),
+				true,
+				BalanceOf::<T>::one().unique_saturated_into(),
+			)?;
+			Self::deposit_event(Event::<T>::AssetRegistered(rid));
+			Ok(())
+		}
+
+		/// Mints Asset into Recipient's Account
+		/// Only Relayers can call it.
+		///
+		/// # Parameters
+		///
+		/// * `origin`: `Asset` owner
+		/// * `destination_add`: Recipient's Account
+		/// * `amount`: Amount to be minted in Recipient's Account
+		/// * `rid`: Resource ID
+		#[allow(clippy::unnecessary_lazy_evaluations)]
+		#[pallet::weight((195_000_000).saturating_add(T::DbWeight::get().writes(2 as Weight)))]
+		pub fn mint_asset(
+			origin: OriginFor<T>,
+			destination_add: Vec<u8>,
+			amount: u128,
+			rid: ResourceId,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let destination_acc = T::AccountId::decode(&mut &destination_add[..])
+				.map_err(|_| Error::<T>::DestinationAddressNotValid)?;
+			ensure!(
+				chainbridge::Pallet::<T>::account_id() == sender,
+				Error::<T>::MinterMustBeRelayer
+			);
```

Comment:
As I said on the sync call this can be moved before decoding and `destination_acc` var definition to not waist time in that if it's not relayer call.